# QC report on mapping, multiplicity and genes detected
```{r load, echo=FALSE, error=FALSE}
prefix = 'qc__'
library(knitr)
opts_chunk$set(message=FALSE, error=FALSE, warning=TRUE, cache=FALSE, auto.dep=TRUE, echo=FALSE, results='hide', fig.path=paste0('figure/', prefix), cache.path=paste0('cache/', prefix))
library(RNASeqPipelineR)
library(MAST)
library(NMF)
library(plyr)
library(data.table)
library(reshape)
library(ggplot2)
library(stringr)

stat.path <- getConfig()[["subdirs"]][["STATS"]]
rsem.path <- getConfig()[["subdirs"]][["RSEM"]]
sheet.path <- file.path(stat.path, 'sample_sheet.rds')
fdat.path <- file.path(rsem.path, 'rsem_fdata.csv')
cdat.path <- sheet.path
tpm.path <- file.path(rsem.path, 'rsem_tpm_matrix.csv')
sheet <- readRDS(sheet.path)

```


## Mapping statistics
```{r mapstats}
## rsemloc defined in RNASeqPipeLineR::mappingReport
## (just a named list of directories with RSEM output)
mapping <- lapply(names(rsemloc), function(nm){
                      mapsum <- RSEMSummarizeMapping(rsemloc[[nm]], log=TRUE, plot=FALSE)
                      mapsum[,dup:=nm]
                      mapsum
           })

mapping <- rbindlist(mapping, fill=TRUE)
setkey(mapping, file)
setkey(sheet, sampleID)
mapping2 <- mapping[unique(sheet),,allow.cartesian=TRUE, nomatch=0]
ggplot(mapping2, aes(x=dup, y=value))+geom_line(aes(group=file), alpha=.2)+ geom_boxplot()+facet_grid(samplePrefix~variable, scale='free_y') + ylab('log10(reads)')+theme_bw()
ggplot(mapping2, aes(x=factor(WellID), y=value, col=variable))+ geom_boxplot()+facet_wrap(~samplePrefix, scale='free_y') + ylab('log10(reads)')+theme_bw()

```

## Multiplicity
```{r multiplicity}
dupstat <- readRDS(file.path(stat.path, 'deduplicateStats.rds'))
pdupstat <- RNASeqPipelineR:::processDupStats(dupstat)
ggplot(pdupstat$mQuantile, aes(x=value))+stat_ecdf() + scale_x_log10()
log10mult <- log10(pdupstat$mQuantile[,value])
dens <- density(log10mult, diff(range(log10mult))/30)
cutot <- data.frame(multiplicity=10^dens$x, y=dens$y, totalreads=cumsum(dens$y*10^(dens$x)))
ggplot(cutot, aes(x=multiplicity, y=totalreads))+geom_line()+scale_x_log10()+scale_y_log10()
ggplot(cutot, aes(x=multiplicity, y=y))+geom_line()+scale_x_log10()
```



```{r readexprmatrix}
fm <- makeSCA(tpm.path, fdat.path, sheet)
```
I found a total of `r nrow(fm)` wells (both bulk and single).  After filtering genes with less than 2% expression, there are `r ncol(fm)` genes.

## Expression statistics
```{r qcMetric, fig.width=8, fig.height=8, eval=FALSE}
ggplot(cData(fm), aes(x=ngeneson))  + geom_text(aes(label=file), size=3)+theme_bw() + facet_wrap(~group)
```

## Heatmap of log2+1 transcripts per million
```{r bulkhm}
aheatmap(exprs(fm))#, annRow=subset(cData(bulk), select=c(cond, cellType)))
```

## Low expressing cells
```{r bulklowngo, results='asis'}
cd <- as.data.table((cData(fm)))
setkey(cd, ngeneson)
kable(cd[1:min(nrow(cd), 40)])
```

## Highly expressed genes
```{r genefreq}
freqs <- freq(fm)
freqs <- data.table(freqs, fData(fm))
freqs[,rankfreq:=rank(-freqs)]
ggplot(freqs[rankfreq<40,], aes(y=gene_symbol, x=freqs))+geom_point()
```
